File: src/main/java/fr/cph/crypto/client/impl/CoinMarketCapClient.java
package fr.cph.crypto.client.impl;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.util.UriComponents;
import org.springframework.web.util.UriComponentsBuilder;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import fr.cph.crypto.client.TickerClient;
import fr.cph.crypto.domain.Currency;
import fr.cph.crypto.domain.Ticker;

@Component
public class CoinMarketCapClient implements TickerClient {

	@Autowired
	private RestTemplate restTemplate;

	@Override
	public Optional<Ticker> getTicker(final Currency currency, final String ticker) {
		return getAllTickers(currency).stream()
				.filter(t -> t.getCurrency().getCode().equals(ticker))
				.findAny();
	}

	@Override
	public List<Ticker> getTickers(final Currency currency, String... tickers) {
		return getTickers(currency, Arrays.asList(tickers));
	}

	@Override
	public List<Ticker> getTickers(final Currency currency, final List<String> tickers) {
		return getAllTickers(currency).stream()
				.filter(ticker -> tickers.contains(ticker.getCurrency().getCode()))
				.collect(Collectors.toList());
	}

	private List<Ticker> getAllTickers(final Currency currency) {
		final UriComponents uriComponents = UriComponentsBuilder.newInstance()
				.scheme("https")
				.host("api.coinmarketcap.com")
				.path("/v1/ticker")
				.queryParam("convert", currency.toString())
				.queryParam("limit", "0")
				.build();

		final Response[] responses = restTemplate.getForObject(uriComponents.toUri(), Response[].class);
		return Arrays.stream(responses)
				.map(response -> TickerMapper.INSTANCE.responseToTicker(response, currency))
				.filter(ticker -> ticker.getCurrency() != Currency.UNKNOWN)
				.collect(Collectors.toList());
	}
}

File: src/main/kotlin/fr/cph/crypto/client/impl/CoinMarketCapClient.kt
package fr.cph.crypto.client.impl

import fr.cph.crypto.client.TickerClient
import fr.cph.crypto.domain.Currency
import fr.cph.crypto.domain.Ticker
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import org.springframework.web.client.RestTemplate
import org.springframework.web.util.UriComponentsBuilder
import java.util.*

@Component
class CoinMarketCapClient @Autowired
constructor(private val restTemplate: RestTemplate) : TickerClient {
    
    override fun getTicker(currency: Currency, ticker: String): Ticker? {
        LOGGER.debug("Search ticker: " + ticker)
        return getAllTickers(currency).firstOrNull { (curr) -> curr.code == ticker }
    }

    override fun getTickers(currency: Currency, vararg tickers: String): List<Ticker> {
        return getTickers(currency, Arrays.asList(*tickers))
    }

    override fun getTickers(currency: Currency, tickers: List<String>): List<Ticker> {
        LOGGER.debug("Search tickers: " + tickers)
        return getAllTickers(currency)
                .filter { (curr) -> tickers.contains(curr.code) }
                .toList()
    }

    private fun getAllTickers(currency: Currency): List<Ticker> {
        val uriComponents = UriComponentsBuilder.newInstance()
                .scheme("https")
                .host("api.coinmarketcap.com")
                .path("/v1/ticker")
                .queryParam("convert", currency.toString())
                .queryParam("limit", "0")
                .build()

        val responses = restTemplate.getForObject(uriComponents.toUri(), Array<Response>::class.java)
        return responses
                .map { response -> TickerMapper.responseToTicker(response, currency) }
                .filter { (curr) -> curr != Currency.UNKNOWN }
                .toList()
    }

    companion object {
        private val LOGGER = LoggerFactory.getLogger(CoinMarketCapClient::class.java)
    }
}
